### PR TITLE
add multiple rate limiter in grpc using a token bucket

### PR DESCRIPTION
**Pull request by reillyse**
# Description

Added in a rate limit for GRPC requests.

We have separate rate limits for each of Workflow, Event and Dispatcher. 
Each api token has it's own rate limit.
The default is 1k / second

I've added a config grpcRateLimit, SERVER_GRPC_RATE_LIMIT

This config is used for each of the rate limits

- [ X] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: internal/services/grpc/middleware/auth.go [hunk 1]
```diff
@@ -41,6 +43,10 @@ func (a *GRPCAuthN) Middleware(ctx context.Context) (context.Context, error) {
 		return nil, forbidden
 	}
 
+	// we want to rate limit by API token
+
+	ctx = context.WithValue(ctx, "rate_limit_token", hashToken(token))
```
Comment (abelanger5) [inline internal/services/grpc/middleware/auth.go]: The token's ID is encoded in the token and should be relatively easy to return from the token package (similar to how we're returning the tenant id). I'd feel more comfortable from a security perspective using the UUID corresponding to the token than setting a hash of the actual token value in the context.

File: internal/services/grpc/middleware/rate_limit.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	fmt.Println("the service is", serviceName)
+
+	rateLimitToken := ctx.Value("rate_limit_token").(string)
+
+	if rateLimitToken == "" {
+		return status.Errorf(codes.Unauthenticated, "no rate limit token found")
+	}
+
+	switch matchServiceName(serviceName) {
+	case "dispatcher":
```
Comment (abelanger5) [inline internal/services/grpc/middleware/rate_limit.go:69]: I don't think we should be rate-limiting the dispatcher, or at least the default rate limit should be higher for the dispatcher. We are not currently seeing issues on this service.

File: pkg/config/server/server.go [hunk 3]
```diff
@@ -97,6 +97,9 @@ type ConfigFileRuntime struct {
 	// GRPCMaxMsgSize is the maximum message size that the grpc server will accept
 	GRPCMaxMsgSize int `mapstructure:"grpcMaxMsgSize" json:"grpcMaxMsgSize,omitempty" default:"4194304"`
 
+	// GRPCRateLimit is the rate limit for the grpc server
```
Comment (abelanger5) [inline pkg/config/server/server.go]: should clarify that this is a per-second, per-engine rate limit

**Review by abelanger5 — CHANGES_REQUESTED** [review]

File: internal/services/grpc/middleware/auth.go (hunk 1 above)
Comment (reillyse) [reply internal/services/grpc/middleware/auth.go]: I now use the ID of the token from the db.

File: internal/services/grpc/middleware/rate_limit.go (hunk 2 above)
Comment (reillyse) [reply internal/services/grpc/middleware/rate_limit.go:69]: ok, set it to 10X the provided rate for the dispatcher - probably best to have some rate limit and a 10X ratio seems safe enough ?

File: pkg/config/server/server.go (hunk 3 above)
Comment (reillyse) [reply pkg/config/server/server.go]: done

File: internal/services/grpc/middleware/rate_limit.go (hunk 2 above)
Comment (abelanger5) [reply internal/services/grpc/middleware/rate_limit.go:69]: sounds good, thanks!

**Review by abelanger5 — APPROVED** [review]

File: internal/services/grpc/middleware/rate_limit.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+		rate:         r,
+		burst:        b,
+		l:            l,
+	}
+}
+
+// Limit is called before each request is processed. It should return an error if rate-limited.
+func (r *HatchetRateLimiter) Limit(ctx context.Context) error {
+
+	fmt.Printf("The rate limit is being checked it is %d \n", int(r.rate))
```
Comment (abelanger5) [inline internal/services/grpc/middleware/rate_limit.go]: Remove debug line

File: pkg/client/client.go [hunk 5]
```diff
@@ -215,9 +217,17 @@ func newFromOpts(opts *ClientOpts) (Client, error) {
 		transportCreds = credentials.NewTLS(opts.tls)
 	}
 
+	retryOpts := []grpc_retry.CallOption{
```
Comment (abelanger5) [inline pkg/client/client.go:225]: Can we case this specifically on `rate limit exceeded` or resource exhausted errors? For a lot of calls, there is no point in retrying

**Review by abelanger5 — CHANGES_REQUESTED** [review]

File: pkg/client/client.go (hunk 5 above)
Comment (reillyse) [reply pkg/client/client.go:225]: I think these are the defaults but I'll make it explicit

Comment (reillyse) [reply pkg/client/client.go:225]: changed to
	grpc_retry.WithCodes(codes.ResourceExhausted, codes.Unavailable),

should help with random network errors but let me know if I should get rid of the Unavailable

**Review by abelanger5 — APPROVED** [review]

**Pull request merged**